pagespeed/kernel/http/response_headers.cc

```diff
@@ -539,17 +539,90 @@ void ResponseHeaders::SetOriginalContentLength(int64 content_length) {
-  return RemoveAllFromSortedArray(&names_to_sanitize[0],
-                                  names_to_sanitize.size());
+  ConstStringStarVector v;
+  StringCompareInsensitive compare;
+  bool changed = false;
+
+  // Sanitize any fields marked as hop-by-hop via the Connection: header
+  if (Lookup(HttpAttributes::kConnection, &v)) {
+    const StringPieceVector& end_to_end =
+        HttpAttributes::SortedEndToEndHeaders();
+    for (int i = 0, n = v.size(); i < n; ++i) {
+      StringPiece val = *v[i];
+      if (val.empty() || StringCaseEqual(val, "keep-alive") ||
+          StringCaseEqual(val, "close") ||
+          StringCaseStartsWith(val, "timeout=") ||
+          StringCaseStartsWith(val, "max=") ||
+          std::binary_search(
+            end_to_end.begin(), end_to_end.end(), val, compare)) {
+        continue;
+      }
+
+      changed = RemoveAll(*v[i]) || changed;
+    }
+  }
+
+  // Remove cookies plus any well-known hop-by-hop headers, which we will never
+  // store in a cache.
+  const StringPieceVector& names_to_sanitize =
+      HttpAttributes::SortedHopByHopHeaders();
+  changed = RemoveAllFromSortedArray(
+      &names_to_sanitize[0], names_to_sanitize.size()) || changed;
+  return changed;
 }
 
 void ResponseHeaders::GetSanitizedProto(HttpResponseHeaders* proto) const {
   Headers<HttpResponseHeaders>::CopyToProto(proto);
   protobuf::RepeatedPtrField<NameValue>* headers = proto->mutable_header();
-  StringPieceVector names_to_sanitize = HttpAttributes::SortedHopByHopHeaders();
+  StringPieceVector more_names;
+
+  // Mark all headers marked as hop-by-hop in "Connection: " for sanitization.
+  for (int i = 0, n = headers->size(); i < n; ++i) {
+    if (StringCaseEqual(headers->Get(i).name(), HttpAttributes::kConnection)) {
+      const StringPieceVector& end_to_end =
+          HttpAttributes::SortedEndToEndHeaders();
+      StringCompareInsensitive compare;
+      StringPieceVector split;
+      SplitStringPieceToVector(headers->Get(i).value(), ",", &split, true);
+      if (split.empty()) {
+        split.push_back(headers->Get(i).value());
+      }
+
+      // Check each value in Connection: val1, val2, ...
+      for (int j = 0, m = split.size(); j < m; ++j) {
+        StringPiece val = split[j];
+        TrimWhitespace(&val);
+        // Skip values that are connection-tokens, empty, or are defined
+        // as being end-to-end.
+        if (val.empty() || StringCaseEqual(val, "keep-alive") ||
+            StringCaseEqual(val, "close") ||
+            StringCaseStartsWith(val, "timeout=") ||
+            StringCaseStartsWith(val, "max=") ||
+            std::binary_search(
+              end_to_end.begin(), end_to_end.end(), val, compare)) {
+          continue;
+        }
+
+        // Find the position at which we should insert to keep the list sorted.
+        std::vector<StringPiece>::iterator up =
+          std::lower_bound(more_names.begin(), more_names.end(), val, compare);
```

Comment:
+2 indent here, and also in argument lists in calls to std::binary_search above.